crypto/x509/x509_vfy.c

```diff
@@ -3224,11 +3225,11 @@ static int build_chain(X509_STORE_CTX *ctx)
             if (!ossl_assert(num == ctx->num_untrusted))
                 goto int_err;
             curr = sk_X509_value(ctx->chain, num - 1);
-            issuer = (self_signed || num > depth) ?
+            issuer = (X509_self_signed(curr, 0) || num > max_depth) ?
```

Comment:
This is change is a consequence of decoupling the previously jointly tracked (curr, self_signed) variables.  In the original code these are scoped function-wide, and retain their value as we iterate the loop.  This avoids having to recompute self_signed more than once for the same certificate (in various cases).

The new code makes these variables internal to the loop, with `curr` and `self_signed` derived from the chain and/or evaluated as needed.  This is perhaps less prone to logic errors in the loop (less state), but is also less efficient.  It is also a mostly cosmetic change in code that is not (at least demonstrably) wrong.

I am not strongly opposed, but I can't entirely say I'm in favour. What do others think?

Reply:
There might be a minor inefficiency if the self-signedess needs to be detected more than once for the same cert, but
* in such a case the result has already been cached in `cert->ex_flags` (in the `EXFLAG_SS` bit posiion) and is used
* the former style of keeping `self_signed` in strong sync also has the potential that is computed needlessly, actually causing less efficiency, whereas now it is computed only whenever needed
* I am not convinced that the value of `self_signed` was actually used beyond the current loop iteration, and if so, this meant a needless extra complication of the already extremely hard-to-understand implementation.

IMO readability and maintainability is much more important here.

Reply:
I agree with @DDvO. The inefficiency is very minor here as the value is cached if the function is called multiple times. The readability and clarity of this critical code is quite important.